File: app-bridge-examples/node-demo/connector.mjs
import http from "node:http";

const host = "127.0.0.1";
const port = Number(process.env.LINGHUN_DEMO_CONNECTOR_PORT ?? 47831);

const capabilities = [
  {
    id: "demo.drawing.describe",
    appId: "demo.drawing",
    title: "Describe Drawing",
    description: "Describes a local drawing.",
    category: "drawing",
    intents: ["describe drawing"],
    keywords: ["drawing", "describe"],
    transport: "http",
    auth: "none",
    permission: "read",
    riskLevel: "low",
    inputSchema: { type: "object", required: ["subject"] },
    outputSchema: { type: "object", required: ["summary"] },
    supportsRollback: false,
    supportsPreview: false,
  },
];

function send(res, status, payload) {
  const body = JSON.stringify(payload);
  res.writeHead(status, {
    "content-type": "application/json; charset=utf-8",
    "content-length": Buffer.byteLength(body),
  });
  res.end(body);
}

async function readJson(req) {
  let body = "";
  req.setEncoding("utf8");
  for await (const chunk of req) body += chunk;
  return body ? JSON.parse(body) : {};
}

const server = http.createServer(async (req, res) => {
  const url = new URL(req.url ?? "/", `http://${host}:${port}`);
  if (req.method === "GET" && url.pathname === "/linghun/capabilities") {
    send(res, 200, { capabilities });
    return;
  }
  if (req.method === "POST" && url.pathname === "/linghun/execute") {
    const request = await readJson(req);
    if (request.capabilityId !== "demo.drawing.describe") {
      send(res, 404, { ok: false, summary: "Capability not found." });
      return;
    }
    const subject = String(request.input?.subject ?? "").trim();
    send(res, 200, {
      ok: true,
      summary: `Described ${subject || "untitled"}.`,
      details: "Demo connector returned bounded details.",
    });
    return;
  }
  send(res, 404, { ok: false, summary: "Not found." });
});

server.listen(port, host, () => {
  console.log(`Demo connector listening on http://${host}:${port}`);
});

File: app-bridge-examples/python-demo/connector.py
from http.server import BaseHTTPRequestHandler, HTTPServer
import json
import os

HOST = "127.0.0.1"
PORT = int(os.environ.get("LINGHUN_DEMO_CONNECTOR_PORT", "47832"))

CAPABILITIES = [
    {
        "id": "demo.notes.summarize",
        "appId": "demo.notes",
        "title": "Summarize Note",
        "description": "Summarizes a local note.",
        "category": "notes",
        "intents": ["summarize note"],
        "keywords": ["note", "summarize"],
        "transport": "http",
        "auth": "none",
        "permission": "read",
        "riskLevel": "low",
        "inputSchema": {"type": "object", "required": ["text"]},
        "outputSchema": {"type": "object", "required": ["summary"]},
        "supportsRollback": False,
        "supportsPreview": False,
    }
]


class Handler(BaseHTTPRequestHandler):
    def _send(self, status, payload):
        body = json.dumps(payload).encode("utf-8")
        self.send_response(status)
        self.send_header("content-type", "application/json; charset=utf-8")
        self.send_header("content-length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_GET(self):
        if self.path == "/linghun/capabilities":
            self._send(200, {"capabilities": CAPABILITIES})
            return
        self._send(404, {"ok": False, "summary": "Not found."})

    def do_POST(self):
        if self.path != "/linghun/execute":
            self._send(404, {"ok": False, "summary": "Not found."})
            return
        raw = self.rfile.read(int(self.headers.get("content-length", "0") or "0"))
        request = json.loads(raw.decode("utf-8") or "{}")
        if request.get("capabilityId") != "demo.notes.summarize":
            self._send(404, {"ok": False, "summary": "Capability not found."})
            return
        text = str(request.get("input", {}).get("text", "")).strip()
        self._send(200, {"ok": True, "summary": f"Summarized {len(text)} chars."})


HTTPServer((HOST, PORT), Handler).serve_forever()
